flaml/nlp/autotransformers.py

```diff
@@ -0,0 +1,901 @@
+        self._set_sample_num_time_budget(custom_num_samples, custom_time_budget, num_sample_time_budget_mode, time_as_grid)
+        scheduler = AutoScheduler.from_scheduler_name(self._scheduler_name)
+
+        self._set_wandb(wandb_key)
+        self.path_utils.make_dir_per_run()
+
+        logger.addHandler(logging.FileHandler(os.path.join(self.path_utils.log_dir_per_run, 'tune.log')))
+        old_level = logger.getEffectiveLevel()
+        self._verbose = verbose
+        if verbose == 0:
+            logger.setLevel(logging.WARNING)
+
+        assert self.path_utils.ckpt_dir_per_run
+
+        start_time = time.time()
+
+        # Documentation on the wandb setting:
+        # There are two ways to initialize wandb in tune.run:
+        # (1) using WandbLoggerCallback, by adding the following argument to tune.run:
+        #     callbacks=[WandbLoggerCallback(
+        #                  project="hpo",
+        #                  api_key = os.environ["WANDB_API_KEY"],
+        #                  group = os.environ["WANDB_RUN_GROUP"],
+        #                  log_config=True)]
+        # (2) using wandb_mixin decorator (the current implementation)
+        # The current implementation uses (2) because (1) has the following bug.
+        # In Ray 1.2, when using WandbLoggerCallback + setting time limit using the time_budget_s argument,
+        # A bug exists which is the previous run will not clear the cache after tune.run returns. After the
+        # later run has already starts, some zombie trials in the previous run remain in the memory and never stop.
+        # This bug can be reproduced by switching to (1) by adding the above callbacks argument and removing the wandb_mixin decorator
+        # https://docs.ray.io/en/master/tune/tutorials/tune-wandb.html
+
+        tune_config = self._search_space_hpo
+        tune_config["wandb"] = {
+                    "project": "hpo",
+                    "group": os.environ["WANDB_RUN_GROUP"],
+                    "reinit": True,
+                    "allow_val_change": True
+                }
+
+        analysis = ray.tune.run(
+            self._objective,
+            metric= self._metric_name,
+            mode = self._metric_mode_name,
+            name = "ray_result",
+            resources_per_trial = resources_per_trial,
+            config= tune_config,
+            local_dir= self.path_utils.ckpt_dir_per_run,
+            num_samples = self._sample_num,
+            time_budget_s= self._time_budget,
+            keep_checkpoints_num = 1,
+            scheduler= scheduler,
+            search_alg= search_algo,
+        )
+        duration = time.time() - start_time
+        self._last_run_duration = duration
+        logger.info("Total running time: {} seconds".format(duration))
+
+        ray.shutdown()
+
+        best_trial = analysis.get_best_trial(scope="all", metric= self._metric_name, mode= self._metric_mode_name)
```

Comment:
space

Reply:
metric_mode_name's choices should be [min, max], I'll change the docstr